### ISSUE TITLE
CluProcessor.parseSentenceWithEisner seems to be crashing

### ISSUE THREAD

**Issue by kwalcock** [clulab/processors#652]
This seems to crash:
```scala
    Utils.initializeDyNet()
    val proc:Processor = new CluProcessor()

    val doc = proc.annotate("contribution , by delegation , to the promotion of the season dry cold ( ssf ) 2019/20 contribution , by speculation , to the enhancement of the ssf 2019/20 podor 27% matam 5% bakel 2% dagana 18% apple earth 10% okra 5% yam 7% but 4% others 26% tomato 10% onion 38% guiers lake 48% valley : 20,992 ha saed/ddac/dsg/bdd , cold dry season 19/20 valley : 20,992 ha saed/ddac/dsg/bdd , cold dry season 2019_2020 1 others : cassava , melon , peanuts , eggplant , watermelon , chili , carrot , cabbage , jaxatu , beans , bissap , banana plantation , fruit trees .")
```

**Comment by kwalcock** [comment]
Specifically
```
[error] java.lang.ArrayIndexOutOfBoundsException: 185
[error]         at org.clulab.dynet.Eisner.$anonfun$generateOutput$2(Eisner.scala:248)
[error]         at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
[error]         at org.clulab.dynet.Eisner.generateOutput(Eisner.scala:244)
[error]         at org.clulab.dynet.Eisner.ensembleParser(Eisner.scala:395)
[error]         at org.clulab.processors.clu.CluProcessor.parseSentenceWithEisner(CluProcessor.scala:458)
[error]         at org.clulab.processors.clu.CluProcessor.$anonfun$parse$1(CluProcessor.scala:773)
[error]         at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
[error]         at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
[error]         at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
[error]         at org.clulab.processors.clu.CluProcessor.parse(CluProcessor.scala:772)
[error]         at org.clulab.processors.clu.CluProcessor.$anonfun$annotate$1(CluProcessor.scala:234)
[error]         at org.clulab.utils.BeforeAndAfter.perform(BeforeAndAfter.scala:10)
[error]         at org.clulab.utils.BeforeAndAfter.perform$(BeforeAndAfter.scala:7)
[error]         at org.clulab.processors.clu.GivenConstEmbeddingsAttachment.perform(CluProcessor.scala:1009)
[error]         at org.clulab.processors.clu.CluProcessor.annotate(CluProcessor.scala:226)
[error]         at org.clulab.processors.Processor.annotate(Processor.scala:65)
[error]         at org.clulab.processors.Processor.annotate$(Processor.scala:62)
[error]         at org.clulab.processors.clu.CluProcessor.annotate(CluProcessor.scala:222)
[error]         at org.clulab.processors.examples.ProcessorExample$.main(ProcessorExample.scala:35)
[error]         at org.clulab.processors.examples.ProcessorExample.main(ProcessorExample.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
```

**Comment by MihaiSurdeanu** [comment]
Thanks Keith! I'll take a look.

**Comment by MihaiSurdeanu** [comment]
I think you're right! There is a check for these bounds in the part of the code that handles the common algorithm, but not in this error handling case...

I think this in line:
```
val depHead = if (relativeHead == 0) 0 else depMod + relativeHead
```
depHead should be bound to be between [0, sentence length)
Thanks!

**Comment by MihaiSurdeanu** [comment]
Thanks @kwalcock !

On the first issue:
```
 startingDependencies(mod + 1)(head + 1).score =
          lambda * startingDependencies(mod + 1)(head + 1).score + (1 - lambda) * topLabelAndScore._2
```
This is fine. The parent for loop goes over `labelTopScores`, which are aligned with `startingDependencies`. That is, we will never "hit" a null cell in this loop.

On the second issue: yes, in the `else` block ("Eisner failed to produce a complete tree; revert to the greedy inference") we do need to check for nulls. In case a null is present, we need to fill in some value. I propose to use: 
0/"root" - for the `if(generateRelativeHeads)` condition, and
-1/"root" - for the `else` condition.
That is, all these words will be headed by the virtual root token in the sentence.

**Comment by kwalcock** [comment]
Addressed with #654.
